Lab 1

Bazalii opened the pull request.

Bazalii committed:
feat: add some declarations of properties and methods

- Add new classes: Person, Product, Shop, ShopManager
- Declare some of their properties and methods

Bazalii committed:
feat: add new classes and some implementations

- Add two Exception classes
- Add Box class for handling products in Shops
- Realise ContainsProduct, RegisterNewProduct and
  first steps in Buy and DeliveryOfProducts methods
  in Shop and ShopManager classes

Bazalii committed:
feat: add new exception class, improve methods

- Add NotInBaseException for products that are not  registered
- Improve Buy method so that it picks all cheapest  pieces of product in several shops
- All fields and methods now use ProductId instead  of ProductName

Bazalii committed:
feat & fix: realise tests and minor fixes

- Add all test for this project
- Some typos, logic and minor bugs fixes

whytrall reviewed (changes requested):
Пока поверхностное ревью, с точки зрения ООП позже гляну

Bazalii committed:
fix: remove setters, correct linq expressions

- Remove unused private setters
- Change logic of some linq expressions

Bazalii committed:
fix: remove all tuples

- Add two new classes to replace Tuples
- Change tests because of usage of new classes
- Remove message with products that are included in delivery but they are not in base

annchous reviewed (changes requested):
Сущности перенести в папку Entities, потому что они - не сервисы.

Bazalii committed:
fix: add a new folder and change class name

- Add Entities folder for corresponding classes
- Rename Order to ProductOrder

Bazalii committed:
fix: add checks and remove blank lines

- Add two new checks: null and negative value
- Add new exceptions for corresponding checks
- Remove several blank lines in Tests

Bazalii committed:
fix: add three new methods

- Make method Buy more readable using three new methods that contain previous logic

Bazalii committed:
fix: remove ShopManager constructor

Bazalii committed:
fix: add usage of Product in Box and ProductOrder

- ProductId in Box -> Product
- ProductName in ProductOrder -> Product
- Add initialization for Box()

Bazalii committed:
fix: correct exceptions and minor fixes

- Remove two own Exception classes
- Add usage of ArgumentException and ArgumentNullException in corresponding checks
- Make Shop address public

Bazalii committed:
fix: correct thrown exception and minor fixes

- Add nameof() and messages for corresponding exceptions
- Rename several methods
- Remove unused method

Bazalii committed:
fix: remove usage of ShopManger in SHop

Bazalii committed:
fix: make Buy method more readable

- Add new methods in ShopManager that contain
  logic from Buy method
- Rename previously added auxiliary methods and
  reduced number of their arguments

Bazalii committed:
fix: several minor changes

- Change method name
- Make Shops and Products private
- Change ProductId to Product argument
  in EnoughProduct method

Bazalii committed:
fix: minor things

- Remove constructor with no parameters
  from Box class
- Remove Console output

Bazalii committed:
feat: add wrap for GetMinimalProductStatistics

- Add new class MinimalProductStatistics
- Add return value to GetMinimalProductStatistics
- Add usage of MinimalProductStatistics object for
  initialization of several local variables
